File: C++/source/Framework/Messaging/LightHouse/LightPulseUnregistrationWorker.cpp
/***************************************************************************
 *   Copyright (C) 20016 Vidyasagara Guntaka                               *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/Messaging/LightHouse/LightPulseUnregistrationWorker.h"
#include "Framework/Messaging/LightHouse/LightPulseRegistrationMessage.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchObjectManager.h"
#include "Framework/Messaging/LightHouse/LightPulseDispatchTypes.h"
#include "Framework/Messaging/LightHouse/LightPulseRegistrationRepository.h"

using WaveNs::LightPulseDispatchObjectManager;

namespace WaveNs
{

LightPulseUnregistrationWorker::LightPulseUnregistrationWorker (LightPulseDispatchObjectManager *pLightPulseDispatchObjectManager)
    : WaveWorker (pLightPulseDispatchObjectManager)
{
    addOperationMap (LIGHT_PULSE_DISPATCH_UNREGISTER_FOR_LIGHT_PULSE, reinterpret_cast<WaveMessageHandler> (&LightPulseUnregistrationWorker::unregisterForLightPulseMessageHandler));
}

LightPulseUnregistrationWorker::~LightPulseUnregistrationWorker ()
{
}

void LightPulseUnregistrationWorker::unregisterForLightPulseMessageHandler (LightPulseRegistrationMessage *pLightPulseRegistrationMessage)
{
    trace (TRACE_LEVEL_DEVEL, "LightPulseUnregistrationWorker::unregisterForLightPulseMessageHandler : Entering ...");

    waveAssert (NULL != pLightPulseRegistrationMessage, __FILE__, __LINE__);

    LightPulseRegistrationRepository *pLightPulseRegistrationRepository = LightPulseRegistrationRepository::getInstance ();

    waveAssert (NULL != pLightPulseRegistrationRepository, __FILE__, __LINE__);

    const string        &lightPulseName = pLightPulseRegistrationMessage->getLightPulseName ();
    const WaveServiceId  waveServiceId  = pLightPulseRegistrationMessage->getSenderServiceCode ();

    pLightPulseRegistrationRepository->removeLightPulseRecepientForLightPulseName (lightPulseName, waveServiceId);

    pLightPulseRegistrationMessage->setCompletionStatus (WAVE_MESSAGE_SUCCESS);
    reply (pLightPulseRegistrationMessage);
}

}

File: Java/source/com/CxWave/Wave/Framework/Messaging/LightHouse/LightPulseUnregistrationWorker.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.Messaging.LightHouse;

import com.CxWave.Wave.Framework.ObjectModel.WaveWorker;
import com.CxWave.Wave.Framework.ObjectModel.WaveWorkerPriority;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.Cardinality;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.OwnerOM;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.WorkerPriority;
import com.CxWave.Wave.Framework.ToolKits.Framework.FrameworkToolKit;
import com.CxWave.Wave.Framework.Type.WaveServiceId;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;

@OwnerOM (om = LightPulseDispatchObjectManager.class)
@Cardinality (1)
@WorkerPriority (WaveWorkerPriority.WAVE_WORKER_PRIORITY_0)
public class LightPulseUnregistrationWorker extends WaveWorker
{
    public LightPulseUnregistrationWorker (final LightPulseDispatchObjectManager lightPulseDispatchObjectManager)
    {
        super (lightPulseDispatchObjectManager);
    }

    private void unregisterForLightPulseMessageHandler (final LightPulseUnregistrationMessage lightPulseUnregistrationMessage)
    {
        trace (TraceLevel.TRACE_LEVEL_DEVEL, "LightPulseUnregistrationWorker.registerForLightPulseMessageHandler : Entering ...");

        waveAssert (null != lightPulseUnregistrationMessage);

        final LightPulseRegistrationRepository lightPulseRegistrationRepository = LightPulseRegistrationRepository.getInstance ();

        waveAssert (null != lightPulseRegistrationRepository);

        final String lightPulseName = lightPulseUnregistrationMessage.getLightPulseName ();
        final WaveServiceId waveServiceId = lightPulseUnregistrationMessage.getSenderServiceCode ();

        lightPulseRegistrationRepository.removeLightPulseRecepientForLightPulseName (lightPulseName, waveServiceId);

        tracePrintf (TraceLevel.TRACE_LEVEL_INFO, true, false, "LightPulseUnregistrationWorker.runegisterForLightPulseMessageHandler : Un-Registered %s for %s", FrameworkToolKit.getServiceNameById (waveServiceId), lightPulseName);

        lightPulseUnregistrationMessage.setCompletionStatus (ResourceId.WAVE_MESSAGE_SUCCESS);
        reply (lightPulseUnregistrationMessage);
    }
}
